### PR TITLE
[Processor: add_cloud_metadata] Use AWS client to get instance metadata and EKS cluster name

### PR DESCRIPTION
**Pull request by tetianakravchenko**
## What does this PR do?

- Use AWS sdk instead of http calls
- Get Identity document to set the same metadata as before
- use EC2 describe tags call to get instance tags, that contains `eks:cluster-name` tag

NOTE: existing issue:
- AWS and Openstack have compatible APIs, so in some cases (if explicitly not defined `providers: ["aws"]`) aws environment is detected as an openstack - it is an existing issue, due to the current implementation how the cloud provider is defined now
related issues:
- https://github.com/elastic/beats/issues/13816
- https://github.com/elastic/beats/issues/31022

## Why is it important?

Add cluster name metadata for EKS clusters, similar to the GKE

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## Screenshots
when used configuration:
```
processors:
      - add_cloud_metadata:
          providers:
            - "aws"
```
Screenshot 2023-04-24 at 10 34 22

### CODE REVIEW

File: libbeat/processors/add_cloud_metadata/provider_aws_ec2.go
Comment (gizas) [inline libbeat/processors/add_cloud_metadata/provider_aws_ec2.go:85]:
```suggestion
		result.err = fmt.Errorf("%w failed loading AWS default configuration", error)
```

File: libbeat/processors/add_cloud_metadata/provider_aws_ec2.go
Comment (gizas) [inline libbeat/processors/add_cloud_metadata/provider_aws_ec2.go:30]: Maybe just import fmt

File: libbeat/processors/add_cloud_metadata/provider_aws_ec2.go
Comment (gizas) [inline libbeat/processors/add_cloud_metadata/provider_aws_ec2.go:93]: Same as above with fmt.errorf

File: libbeat/processors/add_cloud_metadata/provider_aws_ec2.go
Comment (gizas) [inline libbeat/processors/add_cloud_metadata/provider_aws_ec2.go:103]: This can be Warnf level an not debug?

File: libbeat/processors/add_cloud_metadata/provider_aws_ec2.go
Comment (tetianakravchenko) [reply libbeat/processors/add_cloud_metadata/provider_aws_ec2.go]: done - https://github.com/elastic/beats/pull/35182/commits/0241f2270f774f17f9a345d393bfba28219550d8

**Review by gizas — APPROVED** [review]

File: dev-tools/kubernetes/Tiltfile
Comment (kaiyan-sheng) [inline dev-tools/kubernetes/Tiltfile:51]:
```suggestion
        # In order to push to AWS you need to run:
```

File: libbeat/processors/add_cloud_metadata/provider_aws_ec2.go
Comment (kaiyan-sheng) [inline libbeat/processors/add_cloud_metadata/provider_aws_ec2.go:147]: quick question here: since we are only returning `Tags[0]`, does that mean we only have one set of tag key/value returned from the DescribeTags API call? If so, should we check len(tagsResult.Tags) == 1 instead?

**Review by kaiyan-sheng — DISMISSED** [review]
LGTM, just left with one question.

File: libbeat/processors/add_cloud_metadata/provider_aws_ec2.go
Comment (tetianakravchenko) [inline libbeat/processors/add_cloud_metadata/provider_aws_ec2.go:120]: this field is already defined in ecs [orchestrator.cluster.id](https://www.elastic.co/guide/en/ecs/current/ecs-orchestrator.html#field-orchestrator-cluster-id)

File: libbeat/processors/add_cloud_metadata/provider_aws_ec2.go
Comment (tetianakravchenko) [reply libbeat/processors/add_cloud_metadata/provider_aws_ec2.go:147]: yes, there only 1 element is expected, it is basically the same as to run:
```
instance_id=$(curl http://169.254.169.254/latest/meta-data/instance-id)
aws ec2 describe-tags --filter Name=resource-id,Values=$instance_id Name=key,Values=eks:cluster-name --region us-east-2
{
    "Tags": [
        {
            "Key": "eks:cluster-name",
            "ResourceId": "i-020c696ca32297693",
            "ResourceType": "instance",
            "Value": "cluster-name"
        }
    ]
}
```

**Review by gizas — DISMISSED** [review]

File: libbeat/processors/add_cloud_metadata/provider_aws_ec2.go
Comment (tetianakravchenko) [reply libbeat/processors/add_cloud_metadata/provider_aws_ec2.go]: @gizas thank you for pointing it out! From my understanding it should be fine to use it this way, http fetcher uses this definition as well - https://github.com/elastic/beats/blob/98b5cb5a60108be18ba94d4af4dc3af199e64c1c/libbeat/processors/add_cloud_metadata/http_fetcher.go#L94

File: libbeat/processors/add_cloud_metadata/provider_aws_ec2.go
Comment (tetianakravchenko) [reply libbeat/processors/add_cloud_metadata/provider_aws_ec2.go]: @kaiyan-sheng I've changed it in https://github.com/elastic/beats/pull/35182/commits/79070e6ec78a77b72227fc0b7877934227beeccd
I've found a mistake I make in `types.TagDescription` struct, now it return exactly 1 element

**Review by gizas — DISMISSED** [review]

**Review by kaiyan-sheng — APPROVED** [review]

**Comment by tetianakravchenko** [comment]
@pierrehilbert @fearful-symmetry could you please review this PR and help with merging it?

I can't merge it, since I am missing the code owners review

Screenshot 2023-05-15 at 20 24 48

**Comment by tetianakravchenko** [comment]
Hey @elastic/elastic-agent-data-plane, could you help with this PR? I can't merge it because `Waiting on code owner review from elastic/elastic-agent-data-plane.` https://github.com/elastic/beats/pull/35182#issuecomment-1548350175

**Comment by MichaelKatsoulis** [comment]
@tetianakravchenko have you tested that on AWS Fargate ?

**Comment by tetianakravchenko** [comment]
> @tetianakravchenko have you tested that on AWS Fargate ?

@MichaelKatsoulis no, only on the standard EKS setup. Do you think it is a blocker for this PR?

**Comment by MichaelKatsoulis** [comment]
> @MichaelKatsoulis no, only on the standard EKS setup. Do you think it is a blocker for this PR?

Not a blocker, I was just interested on what happens in that case. I would expect to not get the cluster name. 
It could be worthwhile testing to ensure that it doesn't cause anything undesired.

File: libbeat/processors/add_cloud_metadata/provider_aws_ec2.go
Comment (fearful-symmetry) [inline libbeat/processors/add_cloud_metadata/provider_aws_ec2.go:83]: This is an error message, is there a reason the log level is debug? Should be `Errorf`.

**Comment by fearful-symmetry** [comment]
Not sure why github set me as the code owner here, since I don't have a lot of context for the cloud processors. Tried to give a bit of feedback.

File: libbeat/processors/add_cloud_metadata/provider_aws_ec2.go
Comment (gizas) [reply libbeat/processors/add_cloud_metadata/provider_aws_ec2.go]: My comment has to do more with the level error vs debug vs warn. I would say not to use debug as it would hide any messages unless you raise the log level

Comment (tetianakravchenko) [reply libbeat/processors/add_cloud_metadata/provider_aws_ec2.go]: this code is outdated - was replaced to `Warn` in https://github.com/elastic/beats/pull/35182/commits/0241f2270f774f17f9a345d393bfba28219550d8

**Comment by tetianakravchenko** [comment]
> Not sure why github set me as the code owner here, since I don't have a lot of context for the cloud processors. Tried to give a bit of feedback.

@fearful-symmetry I believe it is due to adding new dependency

File: libbeat/processors/add_cloud_metadata/provider_aws_ec2.go
Comment (tetianakravchenko) [reply libbeat/processors/add_cloud_metadata/provider_aws_ec2.go]: `add_cloud_metadata` tries to fetch the metadata from all the available cloud providers, so if running on GCP - `fetchMetadata` for AWS will still run and in this case it might e confising to see an error on GCP regarding the AWS configuration. I've changed it to `Warnf` as it was done before - https://github.com/elastic/beats/pull/35182/commits/19ab15410b8407f67940ae1c79ada52060f93c9b

**Comment by DingGGu** [comment]
This makes fetch IMDS Error on non EKS envs

```json
{"log.level":"warn","@timestamp":"2023-07-26T02:20:11.324Z","log.logger":"add_cloud_metadata","log.origin":{"file.name":"add_cloud_metadata/provider_aws_ec2.go","file.line":102},"message":"error fetching cluster name metadata: error fetching EC2 Tags: operation error EC2: DescribeTags, failed to sign request: failed to retrieve credentials: failed to refresh cached credentials, failed to get nodes EC2 IMDS role credentials, operation error ec2imds: GetMetadata, request canceled, context deadline exceeded.","service.name":"filebeat","ecs.version":"1.6.0"}
{"log.level":"warn","@timestamp":"2023-07-26T02:20:16.470Z","log.logger":"add_cloud_metadata","log.origin":{"file.name":"add_cloud_metadata/provider_aws_ec2.go","file.line":102},"message":"error fetching cluster name metadata: error fetching EC2 Tags: operation error EC2: DescribeTags, exceeded maximum number of attempts, 3, failed to sign request: failed to retrieve credentials: failed to refresh cached credentials, failed to get nodes EC2 IMDS role credentials, operation error ec2imds: GetMetadata, exceeded maximum number of attempts, 3, http response error StatusCode: 500, request to EC2 IMDS failed.","service.name":"filebeat","ecs.version":"1.6.0"}
```

**Pull request merged**
